File: src/workflows.ts
import { proxyActivities, defineSignal, defineQuery, setHandler, condition, log, ActivityFailure } from '@temporalio/workflow';
import type * as activities from './activities';

const gog = proxyActivities<typeof activities>({
  startToCloseTimeout: '30s',
  retry: {
    maximumAttempts: 5,
    initialInterval: '5s',
    backoffCoefficient: 2,
  },
});

const usps = proxyActivities<typeof activities>({
  startToCloseTimeout: '45s',
  retry: {
    maximumAttempts: 5,
    initialInterval: '5s',
    backoffCoefficient: 2,

  },
});

const llm = proxyActivities<typeof activities>({
  startToCloseTimeout: '2m',
  scheduleToCloseTimeout: '15m',
  retry: {
    maximumAttempts: 10,
    initialInterval: '15s',
    backoffCoefficient: 2,
    maximumInterval: '3m',
  },
});

const delivery = proxyActivities<typeof activities>({
  startToCloseTimeout: '15s',
  retry: {
    maximumAttempts: 5,
    initialInterval: '3s',
    backoffCoefficient: 2,
  },
});

const reminder = proxyActivities<typeof activities>({
  startToCloseTimeout: '15s',
  retry: {
    maximumAttempts: 3,
    initialInterval: '3s',
    backoffCoefficient: 2,
  },
});

export const stopRemindersSignal = defineSignal('stopReminders');
export const briefQuery = defineQuery<string | null>('getBrief');

export async function morningBriefWorkflow(): Promise<string> {
  let briefResult: string | null = null;
  let stopped = false;

  setHandler(briefQuery, () => briefResult);
  setHandler(stopRemindersSignal, () => {
    log.info('Received stop signal, cancelling reminders');
    stopped = true;
  });

  log.info('Starting morning brief workflow');

  // Fetch all data in parallel — each retries independently
  const [calendar, emails, uspsScans] = await Promise.all([
    gog.fetchCalendar(),
    gog.fetchEmails(),
    usps.fetchUSPSMailScans(),
  ]);

  log.info('All data fetched, parsing lunch meetings and generating brief');

  // Parse lunch meetings and generate brief in parallel — they're independent
  const [lunchMeetings] = await Promise.all([
    reminder.parseLunchMeetings(calendar),
    (async () => {
      try {
        const brief = await llm.generateBrief({ calendar, emails, uspsScans });
        log.info('Brief generated, sending to Telegram');
        await delivery.sendToTelegram(brief);
        briefResult = brief;
        log.info('Morning brief delivered');
      } catch (err) {
        if (!(err instanceof ActivityFailure)) throw err;
        log.error(`Brief generation failed, skipping: ${err.message}`);
      }
    })(),
  ]);

  if (lunchMeetings.length > 0) {
    log.info(`Found ${lunchMeetings.length} lunch meeting(s), setting up reminders`);
    await sendLunchReminders(lunchMeetings, () => stopped);
  }

  log.info('Morning brief workflow complete');
  return briefResult ?? '';
}

async function sendLunchReminders(meetings: activities.LunchMeeting[], isStopped: () => boolean): Promise<void> {

  for (const meeting of meetings) {
    if (isStopped()) break;

    // Build today's meeting timestamp
    const now = new Date(Date.now());
    const meetingTime = new Date(now);
    meetingTime.setHours(meeting.startHour, meeting.startMinute, 0, 0);
    const meetingMs = meetingTime.getTime();

    // --- 30-minute reminder ---
    const thirtyMinBefore = meetingMs - 30 * 60 * 1000;
    let waitMs = thirtyMinBefore - Date.now();

    if (waitMs > 0) {
      const wasStoppedEarly = await condition(isStopped, waitMs);
      if (wasStoppedEarly) break;
    }

    if (!isStopped() && Date.now() < meetingMs) {
      await delivery.sendToTelegram(
        `⏰ *30 min reminder:* ${meeting.title}\n\n_Reply STOP to cancel reminders._`,
      );
      log.info(`Sent 30-min reminder for: ${meeting.title}`);
    }

    // --- 10-minute reminder ---
    const tenMinBefore = meetingMs - 10 * 60 * 1000;
    waitMs = tenMinBefore - Date.now();

    if (waitMs > 0) {
      const wasStoppedEarly = await condition(isStopped, waitMs);
      if (wasStoppedEarly) break;
    }

    if (!isStopped() && Date.now() < meetingMs) {
      await delivery.sendToTelegram(
        `⏰ *10 min reminder:* ${meeting.title}`,
      );
      log.info(`Sent 10-min reminder for: ${meeting.title}`);
    }
  }
}

File: workflows.py
import asyncio
from datetime import timedelta

from temporalio import workflow
from temporalio.common import RetryPolicy

with workflow.unsafe.imports_passed_through():
    from activities import BriefInput, LunchMeeting


@workflow.defn
class MorningBriefWorkflow:
    def __init__(self) -> None:
        self._brief: str | None = None
        self._stopped = False

    @workflow.query
    def get_brief(self) -> str | None:
        return self._brief

    @workflow.signal
    def stop_reminders(self) -> None:
        workflow.logger.info("Received stop signal, cancelling reminders")
        self._stopped = True

    @workflow.run
    async def run(self) -> str:
        workflow.logger.info("Starting morning brief workflow")

        retry_5 = RetryPolicy(
            maximum_attempts=5,
            initial_interval=timedelta(seconds=5),
            backoff_coefficient=2,
        )
        retry_3 = RetryPolicy(
            maximum_attempts=3,
            initial_interval=timedelta(seconds=3),
            backoff_coefficient=2,
        )

        # Fetch all data in parallel
        calendar_task = workflow.execute_activity(
            "fetch_calendar",
            start_to_close_timeout=timedelta(seconds=30),
            retry_policy=retry_5,
        )
        emails_task = workflow.execute_activity(
            "fetch_emails",
            start_to_close_timeout=timedelta(seconds=30),
            retry_policy=retry_5,
        )
        usps_task = workflow.execute_activity(
            "fetch_usps_mail_scans",
            start_to_close_timeout=timedelta(seconds=45),
            retry_policy=retry_5,
        )

        calendar, emails, usps_scans = await asyncio.gather(
            calendar_task, emails_task, usps_task,
        )

        workflow.logger.info("All data fetched, generating brief")

        brief = await workflow.execute_activity(
            "generate_brief",
            BriefInput(calendar=calendar, emails=emails, usps_scans=usps_scans),
            start_to_close_timeout=timedelta(seconds=60),
            retry_policy=RetryPolicy(
                maximum_attempts=5,
                initial_interval=timedelta(seconds=10),
                backoff_coefficient=2,
            ),
        )

        workflow.logger.info(f"Brief generated:\n{brief}")

        self._brief = brief

        workflow.logger.info("Morning brief complete, checking for lunch meetings")

        lunch_meetings: list[LunchMeeting] = await workflow.execute_activity(
            "parse_lunch_meetings",
            calendar,
            start_to_close_timeout=timedelta(seconds=15),
            retry_policy=retry_3,
        )

        if lunch_meetings:
            workflow.logger.info(f"Found {len(lunch_meetings)} lunch meeting(s), setting up reminders")
            await self._send_lunch_reminders(lunch_meetings)

        workflow.logger.info("Morning brief workflow complete")
        return brief

    async def _send_lunch_reminders(self, meetings: list[LunchMeeting]) -> None:
        delivery_retry = RetryPolicy(
            maximum_attempts=5,
            initial_interval=timedelta(seconds=3),
            backoff_coefficient=2,
        )

        for meeting in meetings:
            if self._stopped:
                break

            now = workflow.now()
            meeting_time = now.replace(
                hour=meeting.start_hour, minute=meeting.start_minute, second=0, microsecond=0,
            )
            meeting_ms = meeting_time.timestamp() * 1000

            # 30-minute reminder
            thirty_min_before = meeting_ms - 30 * 60 * 1000
            wait_ms = thirty_min_before - workflow.now().timestamp() * 1000

            if wait_ms > 0:
                try:
                    await workflow.wait_condition(
                        lambda: self._stopped, timeout=timedelta(milliseconds=wait_ms),
                    )
                except asyncio.TimeoutError:
                    pass  # Timer expired, time to send the reminder
                if self._stopped:
                    break

            if not self._stopped and workflow.now().timestamp() * 1000 < meeting_ms:
                await workflow.execute_activity(
                    "send_to_telegram",
                    f"⏰ *30 min reminder:* {meeting.title}\n\n_Reply STOP to cancel reminders._",
                    start_to_close_timeout=timedelta(seconds=15),
                    retry_policy=delivery_retry,
                )

            # 10-minute reminder
            ten_min_before = meeting_ms - 10 * 60 * 1000
            wait_ms = ten_min_before - workflow.now().timestamp() * 1000

            if wait_ms > 0:
                try:
                    await workflow.wait_condition(
                        lambda: self._stopped, timeout=timedelta(milliseconds=wait_ms),
                    )
                except asyncio.TimeoutError:
                    pass
                if self._stopped:
                    break

            if not self._stopped and workflow.now().timestamp() * 1000 < meeting_ms:
                await workflow.execute_activity(
                    "send_to_telegram",
                    f"⏰ *10 min reminder:* {meeting.title}",
                    start_to_close_timeout=timedelta(seconds=15),
                    retry_policy=delivery_retry,
                )

